README.md
# aroma.cn.eval: Objective Assessment of Copy-number Estimates


## Installation
R package aroma.cn.eval is only available via [GitHub](https://github.com/HenrikBengtsson/aroma.cn.eval) and can be installed in R as:
```r
source('http://callr.org/install#HenrikBengtsson/aroma.cn.eval')
```


## Software quality

| Resource:     | GitHub        | Travis CI        | Appveyor         |
| ------------- | ------------------- | ---------------- | ---------------- |
| _Platforms:_  | _Multiple_          | _Linux_          | _Windows_        |
| R CMD check   |  | <a href="https://travis-ci.org/HenrikBengtsson/aroma.cn.eval"><img src="https://travis-ci.org/HenrikBengtsson/aroma.cn.eval.svg" alt="Build status"></a>    | <a href="https://ci.appveyor.com/project/HenrikBengtsson/aroma-cn-eval"><img src="https://ci.appveyor.com/api/projects/status/github/HenrikBengtsson/aroma.cn.eval" alt="Build status"></a> |
| Test coverage |                     | <a href="https://coveralls.io/r/HenrikBengtsson/aroma.cn.eval"><img src="https://coveralls.io/repos/HenrikBengtsson/aroma.cn.eval/badge.png?branch=develop" alt="Coverage Status"/></a> |                  |


NEWS
Package: aroma.cn.eval
======================

Version: 0.5.4-9000 [2019-01-06]

 * ...
 

Version: 0.5.4 [2019-01-06]

BUG FIXES:

 * aroma.core no longer exports an apply() function.


Version: 0.5.3 [2015-05-04]

SIGNIFICANT CHANGES:

 * Package now requires R (>= 3.1.3) released March 2015. This allows
   us to use BioC (>= 3.0) (October 2014).

SOFTWARE QUALITY:

 * ROBUSTNESS: Forgot to declare one S3 method in NAMESPACE and
   explicitly import R.utils::GenericSummary.


Version: 0.5.2 [2014-03-24]

NEW FEATURES:

 * Now fitRoc() also returns the number of positives and negatives.

 * Added argument 'as' to fitRoc().

 * Now RocCurve() also accepts a list with element 'roc' and 'cuts'.

PERFORMANCE:

 * Minor speedup by replacing rm() calls with NULL assignments.

SOFTWARE QUALITY:

 * ROBUSTNESS: Added validation of arguments to RocCurve().

BUG FIXES:

 * auc() for RocCurve now adds (FPR,TPR):s (0,0) and (1,1) if missing.


Version: 0.5.1 [2014-03-09]

SIGNIFICANT CHANGES:

 * Package requires R (>= 3.0.2) [and Bioc (>= 2.13.0)].
 
SOFTWARE QUALITY:

 * Updated package dependencies.


Version: 0.5.0 [2013-12-12]

NEW FEATURES:

 * Added methods for setting up, parsing and extract CN regions -
  methods that were previously only part of the RSP reports code.

 * DOCUMENTATION: Added vignette 'ROC assessment of detecting a change point'

 * DOCUMENTATION: Added help for several more methods.

SOFTWARE QUALITY:

 * CLEANUP: Dropped function definitions that now are part of the package
  from several internal RSP reports.


Version: 0.4.1 [2013-12-09]

SOFTWARE QUALITY:

 * Updated package dependencies.  Lowered the required version of ROC
  just to get it to build on R-Forge.

 * CLEANUP: Shortened too long (> 100 characters) pathnames of some files.


Version: 0.4.0 [2013-10-28]

SIGNIFICANT CHANGES:

 * Package requires R (>= 2.15.0) [and Bioc (>= 2.10.0)].

SOFTWARE QUALITY:

 * Updated package dependencies.

 * ROBUSTNESS: Now importing only what needs to be imported and
  formally declaring all S3 methods in NAMESPACE.

 * CLEANUP: Now scanRocTpAtFp() calls generic function rowWeightedMeans()
  instead of rowWeightedMeans.matrix().

 * CLEANUP: Package no longer depends on R.cache.  Moved several packages
  from 'Depends' to 'Imports'.

 * CLEANUP: Dropped obsolete usage of autoload().


Version: 0.3.3 [2013-09-23]

PERFORMANCE:

 * normalizeTumorBoost() now uses which() instead of whichVector() of
   'R.utils'.  Before R (< 2.11.0), which() used to be 10x slower than
   whichVector(), but now it's 3x faster.

SOFTWARE QUALITY:

 * Updated package dependencies.


Version: 0.3.2 [2012-04-16]

SOFTWARE QUALITY:

 * Updated package dependencies.


Version: 0.3.1 [2012-03-01]

NEW FEATURES:

 * PREPARING FOR FUTURE: Updated TCGA vignette to also work when
   RawGenericSignals extends a data.frame, i.e. no longer provides reference
   variables.

SOFTWARE QUALITY:

 * CLEANUP: There were some unnecessarily large vignette files.


Version: 0.3.0 [2012-02-26]

NEW FEATURES:

 * Added testSeparation() for SegmentedGenomicSignalsInterface. It used to be
   part of the TCGA vignette.

SOFTWARE QUALITY:

 * CLEANUP: Now the TCGA vignette utilizes toPNG()/toPDF() of R.utils
   (>= 1.11.0) to generate figures.


Version: 0.2.0 [2011-11-29]

SOFTWARE QUALITY:

 * Added a namespace.


Version: 0.1.4 [2011-11-29]

DEPRECATED AND DEFUNCT:

 * blockAvg() has been replaced by colAvgsPerRowSet() in the matrixStats
   package.


Version: 0.1.3 [2011-08-01]

BUG FIXES:

 * WORKAROUND: In order for the package to work with the most recent version of
   R devel, which automatically add namespaces to packages who do not have one,
   we explicitly have specify that this package should get function such as
   cat() from R.utils (instead of 'base').


Version: 0.1.2 [2011-03-18]

DOCUMENTATION:

 * VIGNETTES: Updated the TumorBoost vignettes with more help comments and to
   do more sanity checks and give more informative error messages.


Version: 0.1.1 [2009-11-01]

SOFTWARE QUALITY:

 * Updated the dependencies.  For the record, the ROC package is only needed in
   auc() of the RocCurve class.


Version: 0.1.0 [2009-10-02]

SOFTWARE QUALITY:

 * Updated the dependencies.


Version: 0.0.5 [2009-06-10]

NEW FEATURES:

 * Now all ROC functions are implemented for SegmentedGenomicSignalsInterface.

DEPRECATED AND DEFUNCT:

 * Removed ROC functions for SegmentedCopyNumbers.


Version: 0.0.4 [2009-06-04]

SOFTWARE QUALITY:

 * Removed obsolete dependency on the R.basic package.


Version: 0.0.3 [2009-02-09]

 * ...
 

Version: 0.0.2 [2009-02-03]

NEW FEATURES:

 * Added ...

 * Added internal fitRoc() and findRocTpAtFp() with Rdoc and examples.


Version: 0.0.1 [2009-01-29]

NEW FEATURES:

 * Created by compiling several generations of scripts for evaluating raw
   copy-number estimates, e.g. CNAG, CNAT4, CRMA, CRMAv2, dChip.

 * Created.


R/scanRocTpAtFp.R
###########################################################################/**
# @RdocDefault scanRocTpAtFp
#
# @title "Find the ROC true-positive (TP) rate for a given false-positive (FP) rate"
#
# \description{
#  @get "title".
# }
#
# @synopsis
#
# \arguments{
#   \item{truth}{An NxM @numeric @matrix of length N.}
#   \item{data}{A NxM @numeric @matrix of length N.}
#   \item{fpRate}{A @double in [0,1] specifying the target FP rate.}
#   \item{...}{Additional arguments passed to @see "findRocTpAtFp".}
#   \item{W}{An optional NxM @numeric @matrix specifying weights in [0,1]
#    specific to each data points in \code{data}.}
#   \item{hs}{A @double @vector with elements equal or greater than one.}
#   \item{fit}{...}
#   \item{shifts}{A @double @vector ...}
#   \item{verbose}{See @see "R.utils::Verbose".}
#   \item{.checkArgs}{If @TRUE, arguments are validated, otherwise not.}
# }
#
# \value{
#   Returns a ...
# }
#
# @author
#
# \seealso{
#   @see "findRocTpAtFp".
#   @see "findRocSmoothingForTpAtFp".
# }
#
# @keyword internal
# @keyword utilities
#*/###########################################################################
setMethodS3("scanRocTpAtFp", "default", function(truth, data, fpRate, ..., W=NULL, hs=seq(from=1, to=10, by=0.1), fit=NULL, shifts=0, verbose=FALSE, .checkArgs=TRUE) {
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Validate arguments
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  if (.checkArgs) {
    # Argument 'truth':
    if (!is.matrix(truth))
      throw("Argument 'truth' is not a matrix.");

    # Argument 'data':
    if (!is.matrix(data))
      throw("Argument 'data' is not a matrix.");

    # Argument 'fpRate':
    fpRate <- Arguments$getDouble(fpRate, range=c(0,1));

    # Argument 'hs':
    hs <- Arguments$getDoubles(hs, range=c(1,Inf));
    hs <- unique(hs);
    hs <- sort(hs);

    # Argument 'verbose':
    verbose <- Arguments$getVerbose(verbose);
    if (verbose) {
      pushState(verbose);
      on.exit(popState(verbose));
    }
  }

  verbose && enter(verbose, "Finding TP rates for different amount of smoothing");
  verbose && cat(verbose, "FP rate: ", fpRate);

  dim <- dim(data);
  verbose && cat(verbose, "Data dimensions: ", paste(dim, collapse="x"));
  if (!identical(dim, dim(truth)))
    throw("Internal error: ");

  nbrOfNAs <- sum(is.na(data));
  verbose && printf(verbose, "Number of missing values: %d\n", nbrOfNAs);

  # Calculate the fraction (0 <= c <= 1) of non-missing values
  fractionOKs <- (1 - nbrOfNAs / length(data));
  verbose && printf(verbose, "Fraction of OK observations: %.4f%%\n",
                                                       100*fractionOKs);

  # Record original (truth, data)
  data0 <- data;
  truth0 <- truth;

  if (is.null(fit)) {
    colnames <- c("h", "hApprox", "tpRateEst", "callRate");
    fit <- matrix(NA_real_, nrow=length(hs), ncol=length(colnames));
    fit[,1] <- hs;
    colnames(fit) <- colnames;
  } else {
    hs <- hs[!(hs %in% fit[,1])];
    # Expand fit for these new ones
    t <- matrix(NA_real_, nrow=length(hs), ncol=ncol(fit));
    t[,1] <- hs;
    fit <- rbind(fit, t);
    t <- NULL; # Not needed anymore
    o <- order(fit[,1]);
    fit <- fit[o,];
    o <- NULL; # Not needed anymore
  }

  # Skip already existing ones.
  for (kk in seq_along(hs)) {
    h <- hs[kk];

    verbose && enter(verbose, sprintf("Iteration #%d of %d", kk, length(hs)));
    verbose && printf(verbose, "h: %.4f\n", h);

    if (h == 1) {
      verbose && enter(verbose, "No smoothing of (truth, data)");
      # No smoothing
      hApprox <- h;
      data <- data0;
      truth <- truth0;
      verbose && exit(verbose);
    } else {
      verbose && enter(verbose, "Smoothing (truth, data)");
      data <- truth <- c();
      for (shift in shifts) {
        idxs <- getBlockAverageMap(n=dim[1], h=h, s=shift);
        if (is.null(W)) {
          dataAvg <- colAvgsPerRowSet(data0, S=idxs);
        } else {
          dataAvg <- colAvgsPerRowSet(data0, W=W, S=idxs,
                                      FUN=rowWeightedMeans, tFUN=TRUE);
        }
        data <- rbind(data, dataAvg);
        truth <- rbind(truth, colAvgsPerRowSet(truth0, S=idxs));
      } # for (shift ...)
      hApprox <- attr(idxs, "hApprox");
      verbose && exit(verbose);
    }

    verbose && enter(verbose, "Finding TP rate at FP rate for smoothed (truth, data)");
    # Find TP rate at given FP rate
    fitT <- findRocTpAtFp(truth, data, fpRate=fpRate, ...,
                                  verbose=less(verbose), .checkArgs=FALSE);
    verbose && exit(verbose);

    tpRateEst <- fitT$tpRateEst;
    callRate <- fitT$callRate;
    verbose && cat(verbose, "Estimated TP rate: ", tpRateEst);
    verbose && cat(verbose, "Call rate: ", callRate);

    rr <- which(h == fit[,1]);

    fit[rr,] <- c(h, hApprox, tpRateEst, callRate);
    verbose && print(verbose, fit);

    verbose && exit(verbose);
  } # while(...)

  attr(fit, "fractionOKs") <- fractionOKs;

  verbose && exit(verbose);

  fit;
}) # scanRocTpAtFp()

############################################################################
# HISTORY:
# 2013-12-12
# o DOCUMENTATION: Added help for scanRocTpAtFp().
# 2013-10-28
# o CLEANUP: Now scanRocTpAtFp() calls generic function rowWeightedMeans()
#   instead of rowWeightedMeans.matrix().
# 2013-09-23
# o SPEEDUP/CLEANUP: normalizeTumorBoost() now uses which() instead of
#   whichVector() of 'R.utils'.  Before R (< 2.11.0), which() used to be
#   10x slower than whichVector(), but now it's 3x faster.
# 2009-02-01
# o Renamed from scanTpAtFpLite() to scanRdocTpAtFp().
# 2008-07-25
# o Updated.
# 2007-08-20
# o Added file caching to fitRoc2().
# 2007-08-19
# o Renamed argument 'call' to 'toCall' in fitRoc().
# 2007-04-15
# o Added scanTpAtFp().
# 2007-04-14
# o Added interpolation in findTpAtFp().
# o Removed gc() from fitRoc().
# o Added findTpAtFp() to locate TP rate for given FP rate.
# 2007-03-2x
# o Added fitRoc().
# o Created.
############################################################################
